stock_available_product_expiry/models/product_product.py

```diff
@@ -107,16 +106,35 @@ def _compute_quantities(self):
             product.qty_expired -= product.outgoing_expired_qty
         return res
 
+    def _get_quants(self, quant_domain, quant_out_domain):
```

Comment:
@rousseldenis You must be sure that the original value for `auto_join` on the field is restored even in case of exception....

```python
@api.model
@contextmanager
def _force_auto_join(self, model_field_list):
    """ Force  auto_join for each item (model_name, field_name) in model_field_list
    The original value are restored on exit
    """
    vals = {}
    for model_name, field_name in model_field_list:
        field = self.env[model_name]._fields[field_name]
        vals[field] = field.auto_join
        field.auto_join = True
    try:
        yield
    finally:
        for field, value in vals.iteritems():
            field.auto_join = value

def  _get_quants(self, quand_domain, quant_out_domain):
    stock_quant_obj = self.env['stock.quant']
    with self._force_auto_join([('stock.quant', 'lot_id')('stock.quant', 'reservation_id')]:
        quants = stock_quant_obj.with_context(lang='').read_group(
            quant_domain, ['product_id', 'qty'], ['product_id'])
        quants_out = stock_quant_obj.with_context(lang='').read_group(
            quant_out_domain, ['product_id', 'qty'], ['product_id'])
        return quants, quants_out
```

Reply:
Done